include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the OpenThread TCP API.
+ *
+ */
+
+#ifndef OPENTHREAD_TCP_H_
+#define OPENTHREAD_TCP_H_
+
+#include <openthread/ip6.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-tcp
+ *
+ * @brief
+ *   This module includes functions that control TCP communication.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This a linked buffer structure for use with TCP.
+ *
+ * A single @p otLinkedBuffer structure references an array of bytes in memory,
+ * via @p mData and @p mLength. The @p mNext field is used to form a chain of
+ * @p otLinkedBuffer structures.
+ */
+typedef struct otLinkedBuffer
+{
+    struct otLinkedBuffer *mNext;  ///< Pointer to the next linked buffer in this chain, or NULL if this is the end of the chain
+    const uint8_t* mData;          ///< Pointer to data referenced by this linked buffer
+    uint16_t mLength;              ///< Length of this linked buffer
+} otLinkedBuffer;
+
+struct otTcpActiveSocket;
+typedef struct otTcpActiveSocket otTcpActiveSocket;
+
+/**
+ * This callback informs the application that the TCP 3-way handshake is
+ * complete and that the connection is now established.
+ */
+typedef void (*otTcpConnectDone)(void *aContext, otTcpActiveSocket *aSocket);
```

Comment:
I think that you typically include the context as the first argument, but I'll let @abtink weigh in and I'll make this revision if he's on board with it. I was including it separately because we probably want to keep `otTcpActiveSocket` and `otTcpPassiveSocket` opaque to applications, so they don't have to access their fields directly. We could also provide an `otTcpActiveGetContext` function if we want to preserve that (though I think what we currently have serves the purpose equally well).